### ISSUE TITLE
dealing with all edgecases of merging reads

### ISSUE THREAD

**Issue by elsherbini** [swo/caravan#1]
I have a bunch of ~250bp paired reads, and the region between the primers is only 215 bases. (This is for a Vibrio HSP60 amplicon). Currently the `best_merge_position` function in `merge.py` doesn't allow for this.

**Comment by elsherbini** [comment]
working on a pull request to make `merge` work in this case and the original use case in hopefully some elegant way.

**Comment by swo** [comment]
Like, they're staggered?

--FORWARD
REVERSE--

**Comment by elsherbini** [comment]
like this:

<pre>
forward -------------CATCGACAAAGCGGTTATCGCGGCTGTTGAAGAGCTGAAGAACCTTTCTGTTCCTTGTTCAGACACGAAAGCTATCGCGCAAGTAGGTACTATCTCTGCGAACTCTGATTCGACAGTAGGTAACATCATTGCTGAAGCGATGGAAAAAGTAGGCCGTGATGGCGTTATCACTGTTGAAGAAGGTCAGGCTCTACAAGACGAGCTAGACGTAGTTGAAGGCATG
reverse ATCTTAAGCGCGGCATCGACAAAGCGGTTATCGCGGCTGTTGAAGAGCTGAAGAACCTTTCTGTTCCTTGTTCAGACACGAAAGCTATCGCGCAAGTAGGTACTATCTCTGCGAACTCTGATTCGACAGTAGGTAACATCATTGCTGAAGCGATGGAAAAAGTAGGCCGTGATGGCGTTATCACTGTTGAAGAAGGTCAGGCTCTACAAGACGAGCTAGACGTAGTTG--------
</pre>

The primers are designed to capture a 215 bp fragment, and the reads were longer than that, into the primer on the other side.

**Comment by elsherbini** [comment]
Yes, exactly like you said. I didn't notice I had to click to expand your blockquote.

**Comment by swo** [comment]
Interesting!

A question: Is "size" the useful way to talk about your expectations for staggered reads? I'm used to saying that I should have a 253 bp amplicon. If there's some other way that's useful to look at staggered reads (e.g., by overhang? idk) then maybe it's worth breaking out into another function (merge_stagger or something) that will have a different interface.

A request: If you know how to write tests, could you put an example of what staggered reads looks like and how they should merge into  a new `test/merge.py` (or `test/merge_stagger.py` or whatever)? That would help me know what we're looking at.

A thought: I'm a little wary of this setup because it requires that _all_ the possible product sizes be staggered or not. Maybe this is always the case in practical settings. I'm just nervous that, if I ever have two 150 bp reads and I ask for a 150 +/- 5 bp merged sequence, this setup will break because the 145-149 size products are overlapped and the 151-155 products are staggered.

**Comment by elsherbini** [comment]
fixed by acc221b ?

**Comment by swo** [comment]
Yes, I think so.

Also, you should check out my merge sizer
https://github.com/swo/caravan/blob/master/tools/merge-sizer.py. It's
rough around the edges, but I find this a really easy way to figure out if
my reads are going to merge the way I expect, in the not/staggered position
I expect, at the size I expect.

**Comment by elsherbini** [comment]
Here is the problem at hand, reads can overlap in lots of ways:

```
# size=14
xxxOVERLAP1---
---OVERLAP2yyy
# expected output: xxx + consensus(OVERLAP1, OVERLAP2) + yyy


# size=8
---OVERLAP1xxx
yyyOVERLAP2---
# expected output: consensus(OVERLAP1,OVERLAP2)


# size=8
OVERLAP1
OVERLAP2
# expected output: consensus(OVERLAP1, OVERLAP2)

# size=11
xxxOVERLAP1
---OVERLAP2
# expected output: xxx + consensus(OVERLAP1, OVERLAP2)

# size=11
OVERLAP1---
OVERLAP2yyy
# expected output: consensus(OVERLAP1, OVERLAP2) + yyy

# no overlap, produce error:
# size > len(x) + len(y)
xxx---
---yyy
# expected output: error
```

I'm working on a solution (potentially with the help of stack overflow) to deal with all those cases, hopefully without devolving into the dozen if statement approach.

**Comment by swo** [comment]
I'm confused re: case 2. I would expect that, if staggering is off, you get
`consensus(OVERLAP1, OVERLAP2)`, but if staggering is on, `yyy` +
`consensus` + `xxx`.

**Comment by elsherbini** [comment]
The resulting merged sequence should always start with the beginning of x and end with the end of y.

**Comment by swo** [comment]
OK, so the overhang on staggered reads should always be deleted? (This is
what usearch does:
http://drive5.com/usearch/manual/cmd_fastq_mergepairs.html) Is there ever a
case when you would want to keep the overhangs for staggered reads? (I
suppose I had guessed that, because you often want to keep the overhangs on
non-staggered reads, you might also want to keep them for staggered reads.)

**Comment by elsherbini** [comment]
Yeah, always want to delete the overhang since it is the primer and adapter of the other direction. Didn't realize usearch already implemented this, very nice. No source available though?

**Comment by swo** [comment]
Nope, usearch is emphatically closed source: the full version costs like
$10k.

I guess "overhang" doesn't make sense for non-staggered reads: "staggered"
means "it has an overhang". Then agreed, we should just always toss it!
